sqlair.go

```diff
@@ -183,28 +165,27 @@ func (q *Query) Iter() *Iterator {
 	var rows *sql.Rows
 	var err error
 	var cols []string
-	var close func() error
-	sqlstmt := q.sqlstmt
-	if q.tx != nil {
-		sqlstmt = q.tx.sqltx.Stmt(q.sqlstmt)
-		close = sqlstmt.Close
-	}
 	if q.pq.HasOutputs() {
-		rows, err = sqlstmt.QueryContext(q.ctx, q.pq.Params()...)
+		if q.tx != nil {
```

Comment:
Instead of checking if the query is a transaction here we could save this function as a closure when we create the query object and remove the `tx` field. That way we have all the logic that differentiates transactions and regular queries in one place.

Before the cache was added [there was an interface](https://github.com/canonical/sqlair/pull/81/files#diff-ef99bcd9bee844cd2964fccf754729a6bc3ba08ea4054a5d7a40d82bb7457067L72) `QuerySubstrate` that captured `QueryContext` and `ExectContext`. I think also worked pretty well but if we can save it as a closure we can also get rid of the three fields there that are parameters for Query and Exec.

Reply:
Then we need to create two closures, one for `Query` and one for `Exec` . I remember the `querySubstrate` interface and I thought of doing it for this PR. I thought this way it was simpler and more evident, it is only 3 extra lines and that saves the interface indirection. The same thing happens with closures, you have to find where they were created to see how it is working, in this particular case of few extra lines, I think it is not worth it.

However, if you think it will be clearer by having either closures or an interface we can do that, just wanted to point out the rationale behind my decision.

Reply:
Is there any reason we can't just have one closure called `Run` that is `Query` or `Exec` depending on `pq.HasOutputs`? I would prefer if the code for executing the query was in `Query` itself rather than in `Iter` because those functions are more concerned with how you use the results rather than how the query is executed. To me it feels more natural to keep that in the `Query` function that is different for queries on dbs and transactions.

Reply:
I had a go at implementing it and I actually found something interesting. The only use of the `q.tx` stored in the `Query` outside of the above is to check if the `tx` is done. It seems like the check is not actually needed. The custom `sqlair.ErrTXDone` was added so that if you try and run a query on a closed transaction you get that error and not one saying `"cannot prepare a statement a on closed transaction"` (which wouldn't make sense because as far as the user is concerned they are not preparing a statement).

If  the underlying `sqltx` is closed then it will return the same error anyway, and since we set `sqlair.ErrTXDone = sql.ErrTxDone` the user can still check for a sqlair error and find it.

The check in `Query` still has some value since it stops the `Query` being generated when the `tx` is closed so its good that we have it, though its main reason for existence is gone with the cache.

That being said, it means we can remove the `tx` from the cache. We still need `pq` for the `ScanArgs` function, but I think replacing the query with a closure would work pretty nicely.

I can send you that patch for the code I wrote, remind me in the new year!

Reply:
You are right that the error check is not needed because it fails later with the same error. I have changed the code and uploaded it so that we can discuss about something more concrete. The problem was that even if we move the code into a closure we still need to check whether we have outputs or not:
```go
        rows, result, err = q.run()
	if q.pq.HasOutputs() {
		if err == nil { // if err IS nil
			cols, err = rows.Columns()
		}
	}
```
The ideal scenario would be that the closure does everything needed to run the query and the surrounding code does not care about checking the outputs. We could stuff the columns into the closure and return more things which I am not a fan of.

In general I see the advantages of using the closures approach to decouple how you run it but I can see the code becoming more difficult to read as well.